### PR TITLE
add state view contract

### PR DESCRIPTION
**Pull request by snreynolds**
A lot of offchain services will want to read core state. This is a lens contract we can deploy to expose that to integrators

### CODE REVIEW

File: src/lens/StateView.sol [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+     * @param poolId The ID of the pool.
+     * @param tick The tick to retrieve the bitmap for.
+     * @return tickBitmap The bitmap of the tick.
+     */
+    function getTickBitmap(PoolId poolId, int16 tick) external view returns (uint256 tickBitmap) {
+        return StateLibrary.getTickBitmap(poolManager, poolId, tick);
+    }
+
+    /**
+     * @notice Retrieves the position info without needing to calulcate the `positionId`.
```
Comment (saucepoint) [inline src/lens/StateView.sol:124]:
```suggestion
     * @notice Retrieves the position info without needing to calculate the `positionId`.
```

**Review by saucepoint — APPROVED** [review]

File: src/lens/StateView.sol [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+     * @param poolId The ID of the pool.
+     * @param positionId The ID of the position.
+     * @return liquidity The liquidity of the position.
+     */
+    function getPositionLiquidity(PoolId poolId, bytes32 positionId) external view returns (uint128 liquidity) {
+        return StateLibrary.getPositionLiquidity(poolManager, poolId, positionId);
+    }
+
+    /**
+     * @notice Live calculate the fee growth inside a tick range of a pool
```
Comment (hensha256) [inline src/lens/StateView.sol:170]:
```suggestion
     * @notice Calculate the fee growth inside a tick range of a pool
```

File: src/lens/StateView.sol [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+     * @dev pools[poolId].feeGrowthInside0LastX128 in Position.Info is cached and can become stale. This function will live calculate the feeGrowthInside
```
Comment (hensha256) [inline src/lens/StateView.sol:171]:
```suggestion
     * @dev pools[poolId].feeGrowthInside0LastX128 in Position.Info is cached and can become stale. This function will calculate the up to date feeGrowthInside
```

File: src/lens/StateView.sol [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        external
+        view
+        returns (uint128 liquidity, uint256 feeGrowthInside0LastX128, uint256 feeGrowthInside1LastX128)
+    {
+        return StateLibrary.getPositionInfo(poolManager, poolId, positionId);
+    }
+
+    /**
+     * @notice Retrieves the liquidity of a position.
+     * @dev Corresponds to pools[poolId].positions[positionId].liquidity. A more gas efficient version of getPositionInfo
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/lens/StateView.sol:160]: This is the comment we have written in core haha... I really just pulled them all into here. and it _is_ more gas efficient for retrieiving liquidity which is I think what this comment means...

File: broadcast/DeployStateView.s.sol/11155111/run-latest.json [hunk 5]
```diff
@@ -0,0 +1,53 @@
+{
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply broadcast/DeployStateView.s.sol/11155111/run-latest.json:1]: I kept them both on purpose bc if we redeploy latest will get overwritten and I wanted to keep a trace of what had been deployed by us.

File: src/lens/StateView.sol [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+     * @return tick The current tick of the pool.
+     * @return protocolFee The protocol fee of the pool.
+     * @return lpFee The swap fee of the pool.
+     */
+    function getSlot0(PoolId poolId)
+        external
+        view
+        returns (uint160 sqrtPriceX96, int24 tick, uint24 protocolFee, uint24 lpFee)
+    {
+        return StateLibrary.getSlot0(poolManager, poolId);
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/lens/StateView.sol:29]: Can I do that separately? i'd like this commit to show the deploy broadcast info with the same bytecode 😄

[the remark this answers is not in the document]
Comment (snreynolds) [reply src/lens/StateView.sol:29]: maybe its not that important tho..

File: src/lens/StateView.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/lens/StateView.sol]: https://github.com/Uniswap/v4-core/pull/807

this fixes the comments in core as well

**Review by hensha256 — APPROVED** [review]

**Pull request merged**